src/cpu/flexcpu/simple_dataflow_thread.hh

```diff
@@ -0,0 +1,248 @@
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: Bradley Wang
+ */
+
+#ifndef __CPU_FLEXCPU_SIMPLE_DATAFLOW_THREAD_HH__
+#define __CPU_FLEXCPU_SIMPLE_DATAFLOW_THREAD_HH__
+
+#include <memory>
+#include <set>
+#include <unordered_map>
+#include <vector>
+
+#include "cpu/exec_context.hh"
+#include "cpu/flexcpu/generic_reg.hh"
+#include "cpu/flexcpu/inflight_inst.hh"
+#include "cpu/flexcpu/simple_dataflow_cpu.hh"
+#include "cpu/inst_seq.hh"
+#include "cpu/reg_class.hh"
+#include "cpu/simple_thread.hh"
+#include "mem/request.hh"
+
+
+/**
+ * Instances of this class represent a logical "Core" within the
+ * SimpleDataflowCPU.
+ */
+class SDCPUThread
+{
+  protected:
+    // BEGIN SDCPUThread Internal variables
+
+    /**
+     * We hold a pointer to the CPU which owns this logical thread.
+     */
+    SimpleDataflowCPU* _cpuPtr;
+
+    /**
+     * The SimpleThread class already contains all internal state necessary to
+     * hold onto the state of a basic CPU, demonstrated by its use in many
+     * other CPU models. Until we find a significant need to hold particular
+     * permanent committed state different from a regular CPU in this library,
+     * there is no reason to redefine the wheel.
+     *
+     * The committed state represents the state resulting from the complete
+     * correct execution of all instructions up to the point represented by
+     * numInstsCommitted.
+     */
+    SimpleThread* _committedState;
+
+    /**
+     * We hold onto the InstSeqNum of the last instruction successfully
+     * committed, for ease of checking if older instruction dependencies have
+     * already been satisfied.
+     */
+    InstSeqNum lastCommittedInstNum;
+
+    // BEGIN Speculative state
+
+    /**
+     * We hold a "speculative" PC ahead of the "true" committed PC, in order to
+     * track what it is we're fetching next, prior to instruction issue. The
+     * "true" PC is only advanced during instruction commit.
+     */
+    TheISA::PCState _speculativePC;
```

Comment:
Hmm... I'm thinking we might want to separate out fetching and decoding from the threads. 

Again, let's revisit this after the in-order version is working.